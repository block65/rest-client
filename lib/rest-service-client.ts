import { CustomError, type CustomErrorSerialized } from '@block65/custom-error';
import { Command } from './command.js';
import { hackyConvertDates, resolveHeaders } from './common.js';
import { ServiceError } from './errors.js';
import type { FetcherMethod, ResolvableHeaders } from './fetcher.js';
import type { RuntimeOptions } from './types.js';
import { isPlainObject } from './utils.js';

export interface RestServiceClientConfig {
  headers?: ResolvableHeaders | undefined;
  credentials?: 'include' | 'omit' | 'same-origin' | undefined;
}

export class RestServiceClient<
  ClientInput extends Record<string, unknown> | void = void,
  ClientOutput extends Record<string, unknown> | void = void,
> {
  readonly #config: RestServiceClientConfig;

  readonly #base: URL;

  readonly #fetcher: FetcherMethod;

  async #res<InputType extends ClientInput, OutputType extends ClientOutput>(
    command: Command<InputType, OutputType>,
    runtimeOptions?: RuntimeOptions,
  ) {
    const { method, pathname, /* query, */ body } = command;
    const query = undefined;

    const url = new URL(`.${pathname}`, this.#base);
    url.search = query
      ? new URLSearchParams(
          Object.entries(query).map(([k, v]) => [k, v?.toString() || '']),
        ).toString()
      : '';

    return this.#fetcher({
      url,
      method,
      body,
      headers: await resolveHeaders({
        // ...headers,
        ...this.#config.headers,
      }),
      ...(runtimeOptions?.signal && { signal: runtimeOptions?.signal }),
    });
  }

  public async send<
    InputType extends ClientInput,
    OutputType extends ClientOutput,
  >(
    command: Command<InputType, OutputType>,
    runtimeOptions?: RuntimeOptions,
  ): Promise<OutputType> {
    const res = await this.#res(command, runtimeOptions);

    if (res.status >= 400) {
      if (
        isPlainObject(res.body) &&
        'code' in res.body &&
        'status' in res.body
      ) {
        throw CustomError.fromJSON(
          res.body as unknown as CustomErrorSerialized,
        );
      }
      throw new ServiceError(res.statusText).debug({ res });
    }

    return hackyConvertDates(res.body) as OutputType;
  }

  public async stream<
    InputType extends ClientInput,
    OutputType extends ClientOutput,
  >(command: Command<InputType, OutputType>, runtimeOptions?: RuntimeOptions) {
    const res = await this.#res(command, runtimeOptions);

    if (res.status >= 400) {
      throw new ServiceError(res.statusText).debug({ res });
    }

    if (res.body instanceof ReadableStream) {
      const reader = res.body?.getReader();
      return reader;
    }
    throw new ServiceError('Unstreamable response').debug({ res });
  }

  constructor(
    base: URL,
    fetcher: FetcherMethod,
    config: RestServiceClientConfig = {},
  ) {
    this.#config = config;

    this.#base = base;

    this.#fetcher = fetcher;
  }
}
