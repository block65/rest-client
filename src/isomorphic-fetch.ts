import { parse } from '@hapi/bourne';
import ky, { type Options } from 'ky-universal';
import type { Jsonifiable } from 'type-fest';
import type {
  FetcherMethod,
  FetcherParams,
  FetcherResponse,
} from '../lib/types.js';

export function createIsomorphicFetcher(
  options: Omit<Options, 'method' | 'json' | 'parseJson' | 'signal'> = {},
): FetcherMethod {
  return async (params: FetcherParams) => {
    const { url, method, body, headers, credentials, signal } = params;

    const res = await ky(url, {
      // overridable
      throwHttpErrors: false,
      ...(credentials && { credentials }),

      ...options,

      // combined
      headers: {
        ...options.headers,
        ...headers,
      },

      // not overridable
      method,
      ...(!!body && { json: body }),
      ...(signal && { signal }),
      parseJson: parse,
    });

    const contentType = res.headers.get('content-type');

    // this can be replaced with any other way of detecting streaming
    const streaming = contentType === 'text/event-stream';

    if (!streaming) {
      const isJson = contentType?.includes('/json');

      const responseBody = await (isJson
        ? res.json<Jsonifiable>()
        : res.text());

      return {
        body: responseBody,
        url: new URL(res.url),
        status: res.status,
        statusText: res.statusText,
        ok: res.ok,
      } satisfies FetcherResponse<Jsonifiable | string>;
    }

    return {
      body: res.body,
      url: new URL(res.url),
      status: res.status,
      statusText: res.statusText,
      ok: res.ok,
    } satisfies FetcherResponse<ReadableStream<Uint8Array> | null>;
  };
}
