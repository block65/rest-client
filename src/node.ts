export * from './index.js';

export { createNodeFetchRequestMethod } from './request-methods/node-fetch.js';
export { createGotRequestMethod } from './request-methods/node-got.js';
