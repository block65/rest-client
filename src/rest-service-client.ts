import type { AsyncReturnType } from 'type-fest';
import {
  Config,
  OpenAPI,
  RequestMethod,
  Resolver,
} from './generated/core/OpenAPI.js';

export interface RestServiceClientConfig {
  requestMethod: RequestMethod;
  id: string;
  basePath?: URL | undefined;
  headers?: Record<string, string> | undefined;
  credentials?: 'include' | 'omit' | 'same-origin' | undefined;
  buildId?: string | undefined;
  userAgent?: string | undefined;
  token?: string | Resolver<string> | undefined;
}

export type ApiMethod = (...args: any[]) => Promise<any>;

export class RestServiceClient<T extends ApiMethod = ApiMethod> {
  readonly #config: Config;

  constructor(config: RestServiceClientConfig) {
    this.#config = {
      ...OpenAPI,
      // WITH_CREDENTIALS: !!config.credentials,
      HEADERS: {
        ...config.headers,
        ...(config.userAgent && { 'user-agent': config.userAgent }),
        ...(config.buildId && {
          'x-build-id': `${config.id}/${config.buildId}`,
        }),
      },
      ...(config.basePath && { BASE: config.basePath }),
      VERSION: OpenAPI.VERSION,
      TOKEN: config.token,
      REQUEST: config.requestMethod,
    };
  }

  public async send<M extends T>(
    method: M,
    ...args: Parameters<M>
  ): Promise<AsyncReturnType<M>> {
    return method(
      ...Array.from({ length: method.length - 1 }, (_, idx) => args[idx]),
      this.#config,
    );
  }

  public sendWithAbort<M extends T>(
    method: M,
    ...args: Parameters<M>
  ): [Promise<AsyncReturnType<M>>, AbortController] {
    const controller = new AbortController();

    return [
      method(
        ...Array.from({ length: method.length - 1 }, (_, idx) => args[idx]),
        {
          ...this.#config,
          signal: controller.signal,
        },
      ),
      controller,
    ];
  }

  public sendWithSignal<M extends T>(
    method: M,
    signal: AbortSignal,
    ...args: Parameters<M>
  ): Promise<Awaited<ReturnType<M>>> {
    return method(
      ...Array.from({ length: method.length - 1 }, (_, idx) => args[idx]),
      {
        ...this.#config,
        signal,
      },
    );
  }

  public isAuthenticated() {
    return typeof this.#config.TOKEN !== 'undefined';
  }

  public logOut() {
    delete this.#config.TOKEN;
  }
}
