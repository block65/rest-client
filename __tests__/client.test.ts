import { createServer } from 'node:http';
import { afterAll, beforeAll, describe, expect, test } from '@jest/globals';
import getPort from 'get-port';
import { RestServiceClient } from '../lib/rest-service-client.js';
import { isomorphicFetcher } from '../src/isomorphic-fetch.js';
import { requestListener } from './server.js';

const server = createServer(requestListener);

const port = await getPort();

describe('Client', () => {
  const client = new RestServiceClient(
    new URL(`http://0.0.0.0:${port}`),
    isomorphicFetcher,
    {
      headers: {
        'x-build-id': 'test/123',
        'x-async': () => Promise.resolve('Bearer 1234567890'),
        'x-func': () => 'hello',
      },
    },
  );

  beforeAll(() => {
    server.listen(port);
  });

  test('200 OK!', async () => {
    const response = await client.send((requestMethod, options) =>
      requestMethod(
        {
          method: 'get',
          pathname: '/200',
        },
        options,
      ),
    );

    expect(response).toMatchInlineSnapshot(`
      [
        1,
        2,
        3,
      ]
    `);
  });

  test('404', async () => {
    await expect(
      client.send((requestMethod, options) =>
        requestMethod(
          {
            method: 'get',
            pathname: '/404',
          },

          options,
        ),
      ),
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      '"Unexpected token < in JSON at position 0"',
    );
  });

  test('JSON Error', async () => {
    await expect(
      client.send((requestMethod, options) =>
        requestMethod(
          {
            method: 'get',
            pathname: '/json-error',
          },

          options,
        ),
      ),
    ).rejects.toThrowErrorMatchingInlineSnapshot('"Bad Request"');
  test('Headers', async () => {
    const response = await client.send((requestMethod, options) =>
      requestMethod(
        {
          pathname: '/my-headers',
          method: 'get',
          headers: {
            'x-merged': 'hello',
          },
        },
        options,
      ),
    );

    expect(response).toMatchInlineSnapshot(`
      {
        "accept": "*/*",
        "accept-encoding": "gzip, deflate",
        "accept-language": "*",
        "connection": "keep-alive",
        "host": "redacted",
        "sec-fetch-mode": "cors",
        "user-agent": "undici",
        "x-async": "Bearer 1234567890",
        "x-build-id": "test/123",
        "x-func": "hello",
        "x-merged": "hello",
      }
    `);
  });

  afterAll((done) => {
    server.close(done);
  });
});
