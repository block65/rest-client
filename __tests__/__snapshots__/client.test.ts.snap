// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client 200 OK! 1`] = `
[
  1,
  2,
  3,
]
`;

exports[`Client 404 1`] = `[Error: Not Found]`;

exports[`Client Headers 1`] = `
{
  "accept": "*/*",
  "accept-encoding": "gzip, deflate",
  "accept-language": "*",
  "connection": "keep-alive",
  "host": "redacted",
  "sec-fetch-mode": "cors",
  "user-agent": "node",
  "x-async": "Bearer 1234567890",
  "x-build-id": "test/123",
  "x-func": "hello",
  "x-merged": "hello",
}
`;
